sensors/simsensor_common/simsensor_reader.c

```diff
@@ -0,0 +1,280 @@
+	int err = 0;
+
+	actField = startOfDataSection;
+
+
+	result->type = SENSOR_TYPE_BARO;
+	result->timestamp = timestamp;
+	result->baro.devId = READER_BARO_DEV_ID;
+
+	result->baro.pressure = reader_getFieldLong(actField, &nextField, &err);
+	if (err != 0) {
+		return -1;
+	}
+	actField = nextField;
+
+	result->baro.temp = reader_getFieldLong(actField, &nextField, &err);
+	if (err != 0) {
+		return -1;
+	}
+	actField = nextField;
+
+	return 0;
+}
+
+
+static time_t reader_nextTimeStampGet(simsens_time_t *timeStruct, time_t timestampFromFile)
+{
+	if (timeStruct->first_scenario_time == READER_TIMESTAMP_NOT_SET) {
+		timeStruct->first_scenario_time = timestampFromFile;
+		timeStruct->last_scenario_timestamp = timestampFromFile;
+	}
+	else if (timeStruct->first_scenario_time == timestampFromFile) {
+		timeStruct->loops_cnt++;
+	}
+	else if (timeStruct->last_scenario_timestamp < timestampFromFile) {
+		timeStruct->last_scenario_timestamp = timestampFromFile;
+	}
+
+	return timestampFromFile + timeStruct->loops_cnt * (timeStruct->last_scenario_timestamp - timeStruct->first_scenario_time);
+}
+
+
+int reader_read(simsens_reader_t *rd, event_queue_t *queue)
+{
+	char *lineBuf = NULL, *actField, *nextField;
+	int sensorID, err = 0;
+	size_t bufLen = 0, curLineLen;
+	time_t timestamp, first_timestamp = READER_TIMESTAMP_NOT_SET;
+	sensor_event_t parsed_result;
+
+	while (eventQueue_canEnqueue(queue)) {
+		curLineLen = getline(&lineBuf, &bufLen, rd->scenario_file);
+		if (curLineLen == -1) {
+			if (!feof(rd->scenario_file)) {
+				err = -1;
+				break;
+			}
+
+			if (fseek(rd->scenario_file, 0, SEEK_SET) != 0) {
+				err = -1;
+				break;
+			}
+			continue;
+		}
+
+		if (reader_isLineHeader(lineBuf)) {
+			continue;
+		}
+
+		actField = lineBuf;
+
+		sensorID = reader_getFieldLong(actField, &nextField, &err);
+		if (err < 0) {
+			break;
+		}
+
+		if (sensorID == READER_END_SCENARIO_INDICATOR) {
+			break;
+		}
+
+		if ((sensorID & rd->sensor_types) == 0) {
+			continue;
+		}
+		actField = nextField;
+
+		timestamp = reader_getFieldLong(actField, &nextField, &err);
+		if (err < 0) {
+			break;
+		}
+		actField = nextField;
+
+		timestamp = reader_nextTimeStampGet(&rd->time_struct, timestamp);
+
+		if (first_timestamp == READER_TIMESTAMP_NOT_SET) {
+			first_timestamp = timestamp;
+		}
+
+		if (timestamp - first_timestamp > rd->time_horizon) {
+			/* Go to start of the line */
+			if (fseek(rd->scenario_file, -curLineLen, SEEK_CUR) != 0) {
+				err = -1;
+			}
+			break;
+		}
+
+		switch (sensorID) {
+			case SENSOR_TYPE_ACCEL:
+				if (reader_accelDataParse(actField, timestamp, &parsed_result) < 0) {
+					err = -1;
+				}
+				break;
+			case SENSOR_TYPE_BARO:
+				if (reader_baroDataParse(actField, timestamp, &parsed_result) < 0) {
+					err = -1;
+				}
+				break;
+		}
+
+		if (err < 0) {
+			break;
+		}
+
+		if (eventQueue_enqueue(queue, &parsed_result) < 0) {
+			free(lineBuf);
+			return -1;
+		}
+	}
+
+	free(lineBuf);
```

Comment:
This free suggests that we are doing dynamic allocation at each `read()` call (actually - `getline` does). We cannot escape the dynamic allocation completely, but we can not do it with every read if we store the line (and size of it) pointer in read structure then getline will only reallocate it if necessary.

We can also pre-allocate it to some expected length so that it is clear who allocates it first (`reader_init`/`reader_open`) and who frees it (`reader_done`/`reader_close`)